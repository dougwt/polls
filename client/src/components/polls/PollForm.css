.btn-delete {
  margin-left: 2em;
}
.btn-delete:hover {
  background-color: #ef476f !important;
  color: #fcfcfc !important;
}
.btn-small {
  height: 32.4px;
  line-height: 32.4px;
  font-size: 13px;
}
.red-text {
  color: #ef476f !important;
}
