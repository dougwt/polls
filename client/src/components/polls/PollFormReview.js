import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Button, Row, Icon, ProgressBar } from 'react-materialize';
import PollDetail from './PollDetail';

export class PollFormReview extends Component {
  componentDidMount() {
    this.props.onReset();
  }

  render() {
    return (
      <div>
        <h5>Please confirm your poll</h5>

        <PollDetail formValues={this.props.formValues} disabled={true} />

        {this.renderError() ? this.renderError() : this.renderSpinner()}

        <Row>
          <Button
            node="a"
            className="red btn-back white-text"
            waves="light"
            onClick={this.props.onCancel}
          >
            Back
            <Icon left>keyboard_arrow_left</Icon>
          </Button>

          {this.renderAsyncButton()}
        </Row>
      </div>
    );
  }

  renderAsyncButton() {
    if (this.props.waiting) {
      return (
        <Button
          className="teal btn-next disabled right white-text"
          waves="light"
        >
          Saving...
          <Icon right>create</Icon>
        </Button>
      );
    }

    return (
      <Button
        className="teal btn-next right white-text"
        waves="light"
        onClick={() => this.props.onSave(this.props.formValues)}
      >
        Create
        <Icon right>create</Icon>
      </Button>
    );
  }

  renderError() {
    if (this.props.error) {
      return (
        <Row className="red-text center-align error">
          <strong>Error:</strong> {this.props.error.message}
          {/* <strong>Error:</strong> {this.props.error.response.data.error} */}
        </Row>
      );
    }
  }

  renderSpinner() {
    if (this.props.waiting) {
      return (
        <Row>
          <ProgressBar />
        </Row>
      );
    }

    return <Row>&nbsp;</Row>;
  }
}
PollFormReview.propTypes = {
  error: PropTypes.object,
  formValues: PropTypes.object.isRequired,
  onReset: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  onSave: PropTypes.func.isRequired,
  waiting: PropTypes.bool.isRequired
};

function mapStateToProps(state) {
  return {
    error: state.poll.error,
    formValues: state.form.pollForm.values,
    waiting: state.poll.waiting
  };
}

export default connect(mapStateToProps)(PollFormReview);
