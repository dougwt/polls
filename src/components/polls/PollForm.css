.input-field .prefix.active {
  color: #06bad6;
}

input:not([type]):focus:not([readonly]),
input[type='text']:not(.browser-default):focus:not([readonly]),
input[type='password']:not(.browser-default):focus:not([readonly]),
input[type='email']:not(.browser-default):focus:not([readonly]),
input[type='url']:not(.browser-default):focus:not([readonly]),
input[type='time']:not(.browser-default):focus:not([readonly]),
input[type='date']:not(.browser-default):focus:not([readonly]),
input[type='datetime']:not(.browser-default):focus:not([readonly]),
input[type='datetime-local']:not(.browser-default):focus:not([readonly]),
input[type='tel']:not(.browser-default):focus:not([readonly]),
input[type='number']:not(.browser-default):focus:not([readonly]),
input[type='search']:not(.browser-default):focus:not([readonly]),
textarea.materialize-textarea:focus:not([readonly]) {
  border-bottom: 1px solid #06bad6;
  -webkit-box-shadow: 0 1px 0 0 #06bad6;
  box-shadow: 0 1px 0 0 #06bad6;
}
input.valid:not([type]),
input.valid:not([type]):focus,
input[type='text'].valid:not(.browser-default),
input[type='text'].valid:not(.browser-default):focus,
input[type='password'].valid:not(.browser-default),
input[type='password'].valid:not(.browser-default):focus,
input[type='email'].valid:not(.browser-default),
input[type='email'].valid:not(.browser-default):focus,
input[type='url'].valid:not(.browser-default),
input[type='url'].valid:not(.browser-default):focus,
input[type='time'].valid:not(.browser-default),
input[type='time'].valid:not(.browser-default):focus,
input[type='date'].valid:not(.browser-default),
input[type='date'].valid:not(.browser-default):focus,
input[type='datetime'].valid:not(.browser-default),
input[type='datetime'].valid:not(.browser-default):focus,
input[type='datetime-local'].valid:not(.browser-default),
input[type='datetime-local'].valid:not(.browser-default):focus,
input[type='tel'].valid:not(.browser-default),
input[type='tel'].valid:not(.browser-default):focus,
input[type='number'].valid:not(.browser-default),
input[type='number'].valid:not(.browser-default):focus,
input[type='search'].valid:not(.browser-default),
input[type='search'].valid:not(.browser-default):focus,
textarea.materialize-textarea.valid,
textarea.materialize-textarea.valid:focus,
.select-wrapper.valid > input.select-dropdown {
  border-bottom: 1px solid #06d6a0;
  -webkit-box-shadow: 0 1px 0 0 #06d6a0;
  box-shadow: 0 1px 0 0 #06d6a0;
}
input:not([type]):focus:not([readonly]) + label,
input[type='text']:not(.browser-default):focus:not([readonly]) + label,
input[type='password']:not(.browser-default):focus:not([readonly]) + label,
input[type='email']:not(.browser-default):focus:not([readonly]) + label,
input[type='url']:not(.browser-default):focus:not([readonly]) + label,
input[type='time']:not(.browser-default):focus:not([readonly]) + label,
input[type='date']:not(.browser-default):focus:not([readonly]) + label,
input[type='datetime']:not(.browser-default):focus:not([readonly]) + label,
input[type='datetime-local']:not(.browser-default):focus:not([readonly])
  + label,
input[type='tel']:not(.browser-default):focus:not([readonly]) + label,
input[type='number']:not(.browser-default):focus:not([readonly]) + label,
input[type='search']:not(.browser-default):focus:not([readonly]) + label,
textarea.materialize-textarea:focus:not([readonly]) + label {
  color: #06bad6;
}
.btn-delete {
  margin-left: 2em;
}
.btn-delete:hover {
  background-color: #ef476f !important;
  color: #fcfcfc !important;
}
.btn-add-choice:hover,
.btn-remove-choice:hover {
  background-color: #06bad6 !important;
}
.btn-small {
  height: 32.4px;
  line-height: 32.4px;
  font-size: 13px;
}
.red-text {
  color: #ef476f !important;
}
