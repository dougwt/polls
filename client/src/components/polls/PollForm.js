import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { reduxForm, Field, FieldArray } from 'redux-form';
import { Button, Row, Icon, ProgressBar } from 'react-materialize';
import PollField from './PollField';
import './PollForm.css';

export class PollForm extends Component {
  componentDidMount() {
    if (this.props.onReset) {
      this.props.onReset();
    }
  }
  render() {
    return (
      <div>
        <form onSubmit={this.props.handleSubmit(this.props.onSubmit)}>
          <Row>
            <Field
              label="What question would you like to ask?"
              icon="question_answer"
              name="question"
              component={PollField}
            />
          </Row>

          <Row>
            <FieldArray
              name="choices"
              component={this.renderChoices.bind(this)}
            />
          </Row>

          {this.renderError() ? this.renderError() : this.renderSpinner()}

          <Row>
            <Button
              className="btn btn-back red white-text"
              onClick={this.props.onCancel}
              type="button"
            >
              Cancel
              <Icon left>close</Icon>
            </Button>

            {this.renderAsyncButton()}

            <Button className="btn-next teal right white-text" waves="light">
              Next
              <Icon right>keyboard_arrow_right</Icon>
            </Button>
          </Row>
        </form>
      </div>
    );
  }

  renderAsyncButton() {
    if (!this.props.onDelete) {
      return;
    }

    if (this.props.waiting) {
      return (
        <Button
          className="btn disabled btn-small btn-delete btn-flat btn-red red-text"
          onClick={this.props.onDelete}
          type="button"
        >
          Deleting...
          <Icon left>delete</Icon>
        </Button>
      );
    }

    return (
      <Button
        className="btn btn-small btn-delete btn-flat btn-red red-text"
        onClick={this.props.onDelete}
        type="button"
      >
        Delete Poll
        <Icon left>delete</Icon>
      </Button>
    );
  }

  renderError() {
    if (this.props.error) {
      return (
        <Row className="red-text center-align error">
          <strong>Error:</strong> {this.props.error.message}
        </Row>
      );
    }
  }

  renderSpinner() {
    if (this.props.waiting) {
      return (
        <Row>
          <ProgressBar />
        </Row>
      );
    }

    return <Row>&nbsp;</Row>;
  }

  renderChoiceField(index) {
    const choiceIcons = {
      1: 'looks_one',
      2: 'looks_two',
      3: 'looks_3',
      4: 'looks_4',
      5: 'looks_5',
      6: 'looks_6',
      7: 'looks_7',
      8: 'looks_8',
      9: 'looks_9'
    };

    return (
      <Field
        label={`Choice ${index}`}
        icon={choiceIcons[index]}
        name={`choice_${index}`}
        key={index}
        component={PollField}
      />
    );
  }

  renderChoiceError() {
    if (this.props.error) {
      return <li className="center-align red-text">{this.props.error}</li>;
    }
  }

  renderChoices({ fields, meta: { error } }) {
    return (
      <ul>
        {fields.map((choice, index) => (
          <li key={index}>{this.renderChoiceField(index + 1)}</li>
        ))}
        {this.renderChoiceError(error)}

        <li className="center-align">
          <Button
            flat
            className={
              'btn-add-choice' + (fields.length > 5 ? ' disabled' : '')
            }
            onClick={() => this.addField(fields)}
            node="a"
          >
            <Icon left>add</Icon>
            Add Choice
          </Button>
          <Button
            flat
            className={
              'btn-remove-choice' + (fields.length < 3 ? ' disabled' : '')
            }
            onClick={() => this.removeField(fields)}
            node="a"
          >
            <Icon left>delete</Icon>
            Remove Choice
          </Button>
        </li>
      </ul>
    );
  }

  addField(fields) {
    if (fields.length <= 5) {
      return fields.push();
    }
  }

  removeField(fields) {
    if (fields.length >= 3) {
      return fields.pop();
    }
  }
}
PollForm.propTypes = {
  error: PropTypes.object,
  handleSubmit: PropTypes.func.isRequired,
  initialValues: PropTypes.object,
  onReset: PropTypes.function,
  onCancel: PropTypes.func.isRequired,
  onDelete: PropTypes.func,
  onSubmit: PropTypes.func.isRequired,
  waiting: PropTypes.bool.isRequired
};

function mapStateToProps(state) {
  return {
    error: state.poll.error,
    waiting: state.poll.waiting
  };
}

function validate(values) {
  const errors = {};

  if (!values.question) {
    errors.question = 'You must provide a poll question';
  }

  if (!values.choices || values.choices.length < 2) {
    errors.choices = { _error: 'At least two choices must be entered' };
  } else if (values.choices.length > 6) {
    errors.choices = { _error: 'No more than six choices allowed' };
  }

  for (
    let choice = 1;
    choice <= (values.choices ? values.choices.length : 0);
    choice++
  ) {
    let field = `choice_${choice}`;
    if (!values[field]) {
      errors[field] = `You must provide text for Choice ${choice}`;
      // errors[field] = `You must provide text for all available choices`;
    }
  }

  return errors;
}

export default compose(
  reduxForm({
    validate,
    form: 'pollForm',
    destroyOnUnmount: false,
    enableReinitialize: true
  }),
  connect(mapStateToProps)
)(PollForm);
