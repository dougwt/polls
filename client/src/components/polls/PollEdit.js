import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import requireAuth from '../requireAuth';
import PollForm from './PollForm';
import PollFormReview from './PollFormReview';
import * as actions from '../../actions';

export class PollEdit extends Component {
  constructor(props) {
    super(props);

    this.state = { showReview: false };
  }

  onCancel() {
    this.setState({ showReview: false });
  }

  onDelete() {
    this.props.deletePoll(this.props.pollId, () => {
      this.props.history.push('/polls');
    });
  }

  onSave(formValues) {
    const poll = {
      question: formValues.question,
      choices: formValues.choices.map((choice, i) => {
        return { text: formValues[`choice_${i + 1}`], votes: 0 };
      })
    };

    if (this.props.auth) {
      poll.owner = this.props.auth._id;
    }
    this.props.editPoll(this.props.pollId, poll, () => {
      this.props.history.push('/polls');
    });
  }

  onSubmit() {
    this.setState({ showReview: true });
  }

  renderContent() {
    if (this.state.showReview) {
      return (
        <PollFormReview
          onCancel={this.onCancel.bind(this)}
          onSave={this.onSave.bind(this)}
        />
      );
    }

    return (
      <PollForm
        onCancel={this.props.onCancel.bind(this)}
        onDelete={this.onDelete.bind(this)}
        onSubmit={this.onSubmit.bind(this)}
        initialValues={this.props.formValues}
      />
    );
  }

  render() {
    return this.renderContent();
  }
}
PollEdit.propTypes = {
  auth: PropTypes.oneOfType([PropTypes.object, PropTypes.bool]).isRequired,
  deletePoll: PropTypes.func.isRequired,
  editPoll: PropTypes.func.isRequired,
  formValues: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  onCancel: PropTypes.func.isRequired,
  polls: PropTypes.object,
  pollId: PropTypes.string
};

function mapStateToProps(state) {
  return {
    auth: state.auth
  };
}

export default compose(
  connect(mapStateToProps, actions),
  requireAuth,
  withRouter
)(PollEdit);
