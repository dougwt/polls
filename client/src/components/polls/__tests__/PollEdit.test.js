import React from 'react';
import { shallow } from 'enzyme';
import { PollEdit } from '../PollEdit';
import PollForm from '../PollForm';
import PollFormReview from '../PollFormReview';

describe('PollEdit', () => {
  let props = {};

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<PollEdit {...props} />);
  });

  it('renders properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('when the page is initially loaded', () => {
    it('contains the proper component state', () => {
      expect(wrapper.state('showReview')).toEqual(false);
    });

    it('shows the PollForm component', () => {
      expect(wrapper.find(PollForm).length).toEqual(1);
    });

    it('does not show the PollFormReview component', () => {
      expect(wrapper.find(PollFormReview).length).toEqual(0);
    });
  });

  describe('when the page has been submitted for review', () => {
    beforeEach(() => {
      wrapper.setState({ showReview: true });
    });

    it('shows the PollFormReview component', () => {
      expect(wrapper.find(PollFormReview).length).toEqual(1);
    });

    it('does not show the PollForm component', () => {
      expect(wrapper.find(PollForm).length).toEqual(0);
    });
  });
});
