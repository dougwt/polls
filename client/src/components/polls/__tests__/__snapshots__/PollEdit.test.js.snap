// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollEdit renders properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PollEdit
    auth={true}
    deletePoll={[Function]}
    editPoll={[Function]}
    formValues={Object {}}
    history={Object {}}
    onCancel={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "initialValues": Object {},
      "onCancel": [Function],
      "onDelete": [Function],
      "onReset": undefined,
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "initialValues": Object {},
        "onCancel": [Function],
        "onDelete": [Function],
        "onReset": undefined,
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
